cli/internal/providers/datacatalog/state/category.go

```diff
@@ -0,0 +1,138 @@
+package state
+
+import (
+	"time"
+
+	catalog "github.com/rudderlabs/rudder-iac/api/client/catalog"
+	"github.com/rudderlabs/rudder-iac/cli/internal/syncer/resources"
+)
+
+// CategoryArgs holds the necessary information to create a category
+type CategoryArgs struct {
+	Name string
+}
+
+func (args *CategoryArgs) ToResourceData() resources.ResourceData {
+	return resources.ResourceData{
+		"name": args.Name,
+	}
+}
+
+func (args *CategoryArgs) FromResourceData(from resources.ResourceData) {
+	args.Name = MustString(from, "name")
+}
+
+// CategoryState represents the full state of a category
+type CategoryState struct {
+	CategoryArgs
+	ID          string
+	Name        string
+	WorkspaceID string
+	CreatedAt   string
+	UpdatedAt   string
+}
+
+func (c *CategoryState) ToResourceData() resources.ResourceData {
+	return resources.ResourceData{
+		"id":           c.ID,
+		"name":         c.Name,
+		"workspaceId":  c.WorkspaceID,
+		"createdAt":    c.CreatedAt,
+		"updatedAt":    c.UpdatedAt,
+		"categoryArgs": map[string]interface{}(c.CategoryArgs.ToResourceData()),
+	}
+}
+
+func (c *CategoryState) FromResourceData(from resources.ResourceData) {
+	c.ID = MustString(from, "id")
+	c.Name = MustString(from, "name")
+	c.WorkspaceID = MustString(from, "workspaceId")
+	c.CreatedAt = MustString(from, "createdAt")
+	c.UpdatedAt = MustString(from, "updatedAt")
+	c.CategoryArgs.FromResourceData(resources.ResourceData(
+		MustMapStringInterface(from, "categoryArgs"),
+	))
+}
+
+// CategoryStateFromAPI converts API Category to CategoryState
+func CategoryStateFromAPI(category *catalog.Category) *CategoryState {
+	return &CategoryState{
+		CategoryArgs: CategoryArgs{
+			Name: category.Name,
+		},
+		ID:          category.ID,
+		Name:        category.Name,
+		WorkspaceID: category.WorkspaceID,
+		CreatedAt:   category.CreatedAt.Format(time.RFC3339),
+		UpdatedAt:   category.UpdatedAt.Format(time.RFC3339),
+	}
+}
+
+// CategoryStateToAPI converts CategoryState to API Category
+func CategoryStateToAPI(state *CategoryState) *catalog.Category {
+	createdAt, _ := time.Parse(time.RFC3339, state.CreatedAt)
+	updatedAt, _ := time.Parse(time.RFC3339, state.UpdatedAt)
+
+	return &catalog.Category{
+		ID:          state.ID,
+		Name:        state.Name,
+		WorkspaceID: state.WorkspaceID,
+		CreatedAt:   createdAt,
+		UpdatedAt:   updatedAt,
+	}
+}
+
+// CategoryStateCollection represents a collection of category states
+type CategoryStateCollection []*CategoryState
+
+// GetByID returns a category by its ID
+func (collection CategoryStateCollection) GetByID(id string) *CategoryState {
```

Comment:
I'm not sure why we need these functions to get resources from state. State is a generic representation that providers have to generate in order to update. Reading is done by the framework in a generic way, I'm not sure there is any need to fetch individual resources from state, or modify the state collection in any way.